test/indexed_test.dart

```diff
@@ -0,0 +1,199 @@
+
+import '../lib/realm.dart';
+
+part 'indexed_test.g.dart';
+
+// IMPORTANT: Don't import our own test.dart here. It will break AOT compilation!
+
+// Realm models cannot extend other classes, but they can implement interfaces.
+class Base {
+  late int anInt;
+  late bool aBool;
+  late String string;
+  late DateTime timestamp;
+  late ObjectId objectId;
+  late Uuid uuid;
+}
+
+@RealmModel()
+class _WithIndexes implements Base {
+  @override
+  @Indexed()
+  late int anInt;
+
+  @override
+  @Indexed()
+  late bool aBool;
+
+  @override
+  @Indexed()
+  late String string;
+
+  @override
+  @Indexed()
+  late DateTime timestamp;
+
+  @override
+  @Indexed()
+  late ObjectId objectId;
+
+  @override
+  @Indexed()
+  late Uuid uuid;
+}
+
+@RealmModel()
+class _NoIndexes implements Base {
+  @override
+  late int anInt;
+
+  @override
+  late bool aBool;
+
+  @override
+  late String string;
+
+  @override
+  late DateTime timestamp;
+
+  @override
+  late ObjectId objectId;
+
+  @override
+  late Uuid uuid;
+}
+
+typedef QueryBuilder<T, U> = RealmResults<T> Function(U value);
+
+Future<void> main([List<String>? args]) async {
+  test('Indexed faster', () {
+    final config = Configuration.local([WithIndexes.schema, NoIndexes.schema]);
+    Realm.deleteRealm(config.path);
+    final realm = Realm(config);
+
+    const max = 100000;
+    final allIndexed = realm.all<WithIndexes>();
+    final allNotIndexed = realm.all<NoIndexes>();
+
+    expect(realm.all<WithIndexes>().length, 0);
+
+    intFactory(int i) => i.hashCode;
+    boolFactory(int i) => i % 2 == 0;
+    stringFactory(int i) => '${i.hashCode} $i';
+    timestampFactory(int i) => DateTime.fromMillisecondsSinceEpoch(i.hashCode);
+    objectIdFactory(int i) => ObjectId.fromValues(i.hashCode * 1000000, i.hashCode, i);
+    uuidFactory(int i) => Uuid.fromBytes(Uint8List(16).buffer..asByteData().setInt64(0, i.hashCode));
+
+    final indexed = List.generate(
+      max,
+      (i) => WithIndexes(
+        intFactory(i),
+        boolFactory(i),
+        stringFactory(i),
+        timestampFactory(i),
+        objectIdFactory(i),
+        uuidFactory(i),
+      ),
+    );
+    realm.write(() => realm.addAll(indexed));
+
+    expect(allIndexed.length, max);
+
+    final notIndexed = List.generate(
+      max,
+      (i) => NoIndexes(
+        intFactory(i),
+        boolFactory(i),
+        stringFactory(i),
+        timestampFactory(i),
+        objectIdFactory(i),
+        uuidFactory(i),
+      ),
+    );
+    realm.write(() => realm.addAll(notIndexed));
+
+    expect(allNotIndexed.length, max);
+
+    // Inefficient, but fast enough for this test
+    final searchOrder = (List.generate(max, (i) => i)..shuffle(Random(42))).take(1000).toList();
+
+    QueryBuilder<T, U> builder<T extends RealmObject, U extends Object>(RealmResults<T> results, String fieldName) {
+      return (U value) => results.query('$fieldName == \$0', [value]);
+    }
+
+    @pragma('vm:no-interrupts')
```

Comment:
Why do we need this. Shouldn't we be testing in real world scenarios. If this is needed perhaps it makes sense to relax the expectation a bit instead of trying too hard to hit the current ones

Reply:
Are you. asking about the `@pragma(vm:no-interrupts)`? Then https://mrale.ph/blog/2021/01/21/microbenchmarking-dart-part-1.html is a good place start.

Since this it a timing call, I rather not have the GC disturb us.

Reply:
Yes I know this post. I am asking about that yes.

Reply:
Why not relax the test expectaions instead of doing it like this. Hence we will be capturing and expecting on a real world usage.

Reply:
Why are you worried about this pragma? I don't understand.

I have relaxed the expected speedup. If this test ever fails on CI, something is really amiss.

Reply:
It's just not running the test in a real world usage. I don't expect many will use that in their code.

Reply:
This is not a real world usage. I'm timing indexed vs non-indexed fields to ensure we see a difference in performance when doing `field == $0` queries, ie. that we didn't somehow forget to communicate the index to core or whatever, or that the index handling in core regressed.

For that to be less brittle, I would like to avoid GC while it runs, hence `@pragma(vm:no-interrupts)`. It is only applied to the `measureSpeed` function - I fail to see why that is controversial.

Reply:
ok resolving, I feel not worth to discuss more